### PR TITLE
Refactor VirtualThread synchronization design

### PR DESCRIPTION
**Pull request by fengxue-IS**
Removes use of `liveVirtualThreadList/liveVirtualThreadListMutex` in the VM and rely on GC's Continuation list.
This design uses an atomic based RW lock with busy waiting for vthread access during mounted/unmount transition.
Global access in JVMTI is a slow path and done through Exclusive VM Access which doesn't require additional locking.

mutual dependency: 
    jdk19: https://github.com/ibmruntimes/openj9-openjdk-jdk19/pull/83
    jdk20: https://github.com/ibmruntimes/openj9-openjdk-jdk20/pull/32
    jdknext: https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/574

Fixes: #16728 

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

### CODE REVIEW

**Comment by fengxue-IS** [comment]
FYI @babsingh @tajila @gacholio

File: runtime/gc_base/GCExtensions.hpp
Comment (amicic) [inline runtime/gc_base/GCExtensions.hpp:359]: From an in-person discussion, I understand that the last unmount (that would normally release native resource)  is not really guarantied as previously believed (is this really a consequence of this redesign, or it was true all along?), so GC pruning pass is still needed. 

This is unfortunate, since this list contains all Continuation objects allocated since previous GC, hence mostly dead by next GC and since most of VTs realistically end up doing last unmount. Doing this pass every GC for those exception cases is not good... Pauses for low pause GCs (like Concurrent Scavenger) may noticeably increase.

We will probably need to optimize the list further. For example, maintain only Continuations that have not finished, by removing them from the list during last unmount.  Would require the list to be double linked. Contention should not be much of any issue since we use multiple lists. @LinHu2016  

Some more advanced things may be considered as well with better understanding under which circumstances VT/continuation can die without doing last unmount (which understanding I don't have). But, for example, if it's because Scheduler died and all its active VTs, we should just process those.

**Comment by fengxue-IS** [comment]
**Update**
1. Per discussion with @babsingh, will merge design concept from #16846 to use EVMA instead of a RW-lock for controlling the global vthread access. This should completely remove the mount/unmount contention seen in stress testing.

2. Refactor getStackTraceImpl to redo the check after calling `haltThreadForInspection` so it will not require using `VirtualThread.inspectorCount` for sync

3. Dynamically set `VirtualThread.notifyJvmtiEvents` to remove all dependency on the `notifyJvmtiMount/Unmount[Begin/End]` API if JVMTI is not used.

File: runtime/gc_base/GCExtensions.hpp
Comment (amicic) [reply runtime/gc_base/GCExtensions.hpp:359]: To answer my own question (after another discussion with @fengxue-IS and @babsingh), seems like the scenario of VTs dying before doing last unmount was possible even in current workloads that we use, but since the VT list served as strong root, and for a potential VT being added on first mount, dying but never removed, it would  be kept alive forever. Not only GC would not be able to see it as a dead object (with dangling stack) and trigger a verification assert, but this scenario would also cause leak in both java heap and native memory.

Now that the VT list is removed, such a VT can really die and the assert triggers, hence enabling  GC list processing must be done as a part of this change.

**Comment by gacholio** [comment]
> 1. use EVMA instead of a RW-lock for controlling the global vthread access

That's fine for the JVMTI code but will probably be unuseably slow for any hot code path. EVMA is an extremely expensive operation.

**Comment by fengxue-IS** [comment]
> > 1. use EVMA instead of a RW-lock for controlling the global vthread access
> 
> That's fine for the JVMTI code but will probably be unuseably slow for any hot code path. EVMA is an extremely expensive operation.

Currently JVMTI suspend/resume all is the only case where the RW lock is using writer access, so I assume by switching to EMVA will just remove the need for any locking for non-jvmti path

File: runtime/oti/VMHelpers.hpp
Comment (tajila) [inline runtime/oti/VMHelpers.hpp:2175]: why not just call `VM_AtomicSupport::add`?

File: runtime/oti/vm_api.h
Comment (tajila) [inline runtime/oti/vm_api.h:4415]: minor: but I think all these functions should be implemented in the same file. ContinuationHelpers makes sense to me

File: runtime/oti/VMHelpers.hpp
Comment (tajila) [reply runtime/oti/VMHelpers.hpp:2175]: oh, I guess it can be racing with an inspector

**Comment by fengxue-IS** [comment]
updated to the EVMA design, perf result 7x better on nocompressedrefs, 20% perf better on compressedrefs

File: runtime/vm/ContinuationHelpers.cpp
Comment (gacholio) [inline runtime/vm/ContinuationHelpers.cpp:389]: The assertion would make more sense after the first read of the inspector count.

File: runtime/jvmti/jvmtiThread.c
Comment (babsingh) [inline runtime/jvmti/jvmtiThread.c:1348]:
```suggestion
				/* Resume the virtual thread. */
				if (NULL != targetThread) {
```

File: runtime/vm/ContinuationHelpers.cpp
Comment (babsingh) [inline runtime/vm/ContinuationHelpers.cpp:85]:
```suggestion
	/* Default inspectorCount value is -1, and it indicates transition. */
```

**Review by babsingh — CHANGES_REQUESTED** [review]
first pass, will review functionality in more detail later.

we should also add a function description for all the new callbacks. it will make them easier to understand and quicker to verify their behaviour.

File: runtime/compiler/control/HookedByTheJit.cpp
Comment (fengxue-IS) [reply runtime/compiler/control/HookedByTheJit.cpp:6828]: no since what is previously walked during scanning cannot be placed back on the buffer

**Comment by fengxue-IS** [comment]
Adressed all comments, @babsingh can you take another look

**Review by babsingh — APPROVED** [review]
Approving, functionally the code LGTM. @0xdaryl for reviewing the JIT code.

Minor nitpicks:
- The PR description should reflect the five commits.
- `Fix vthread suspend order` exceeds the 72 char limit.
- Justification or the reason for actions such as `Move freeContinuation to yieldContinuationImpl` is missing. Such critical details need to be documented.
- Also, the potential perf issue with the spin-only `inspectorCount` solution should be documented. It does not need to be addressed since JVMTI perf is not critical. But, there should be a note or an issue which highlights the CPU hog issue, which may happen in the worst case scenario.

**Comment by babsingh** [comment]
@fengxue-IS I do remember seeing your personal builds. You should post a link here for other reviewers.

**Comment by fengxue-IS** [comment]
[Personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/15985/) passed with https://github.com/ibmruntimes/openj9-openjdk-jdk19/pull/83 (failures are not related to virtual thread

File: runtime/gc_base/GCExtensions.hpp
Comment (gacholio) [reply runtime/gc_base/GCExtensions.hpp:363]: Is the continuation list actually optional now that we're using it instead of the vthread list?

File: runtime/vm/BytecodeInterpreter.hpp
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp:5283]: Is this going to cause any issues with the GC call below (i.e. does the GC call need the native struct for any reason)?

File: runtime/vm/ContinuationHelpers.cpp
Comment (gacholio) [inline runtime/vm/ContinuationHelpers.cpp:388]: Should there be a yield here?

Comment (gacholio) [reply runtime/vm/ContinuationHelpers.cpp:388]: Same comment on the spin above.

**Comment by gacholio** [comment]
There's still a lot of spinning here, so CPU use will need to be measured. For many cloud customers, this matters more than raw performance.

File: runtime/vm/BytecodeInterpreter.hpp
Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: Also missing space after if.

File: runtime/vm/ContinuationHelpers.cpp
Comment (gacholio) [reply runtime/vm/ContinuationHelpers.cpp:388]: This one might not matter as it's a strict decrement operation that can't be interrupted by inspection. A similar argument can be made for the one above.

File: runtime/gc_base/GCExtensions.hpp
Comment (fengxue-IS) [reply runtime/gc_base/GCExtensions.hpp:363]: With this change, continuation list must be enabled to correctly perform JIT codecache clean up and JVMTI operations

File: runtime/vm/ContinuationHelpers.cpp
Comment (fengxue-IS) [reply runtime/vm/ContinuationHelpers.cpp:388]: Given the assert that `vthreadInspectorCount > 0` before the loop, this can only fail if another inspector giveup access and decrement the count at sametime. I don't see yield would help much with this type of contention

File: runtime/vm/BytecodeInterpreter.hpp
Comment (fengxue-IS) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: Their are some related changes GC's concurrent changeset #16713 . From what I am aware, this doesn't impact GC call as they will verify the vmRef pointer before scanning.
@LinHu2016 Can you provide more insight?

Comment (LinHu2016) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: For last unmount continuation, GC should not call postUnmountContinuation for the compensating scan(there is an update in https://github.com/eclipse-openj9/openj9/pull/16713), anyway there is a protection checking vmRef pointer before scanning, so it should not cause any issues for the GC scan after  freeContinuation(), but there is a race case between unmounting and concurrent scanning, which we also fixed in https://github.com/eclipse-openj9/openj9/pull/16713, above the change would neither help the case nor cause more issues.

**Comment by gacholio** [comment]
@tajila Am I overly worried about the spinning/CPU use? The spins occur mainly in JVMTI and `getStackTraceImpl`, which aren't really critical paths.

File: runtime/vm/BytecodeInterpreter.hpp
Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: OK, please still fix the space.

**Comment by tajila** [comment]
> @tajila Am I overly worried about the spinning/CPU use? The spins occur mainly in JVMTI and getStackTraceImpl, which aren't really critical paths.

I agree that the paths were spinning is potentially going to be an issue arent really critical paths. I do think that we can optimize the `acquireVThreadInspector` and `enterVThreadTransitionCritical` paths to block if spinning occurs a lot. I also think that acquire/release VMAccess on every iteration we spin is a bit excessive. But overall I think this is the right approach and we can optimize those two functions in the near future.

File: runtime/jcl/common/thread.cpp
Comment (tajila) [reply runtime/jcl/common/thread.cpp:344]: Or maybe parameterize `acquireVThreadInspector` with a noSpin param. It will make it simpler in the future of we need to update this logic to only have it in one place

File: runtime/oti/VMHelpers.hpp
Comment (gacholio) [reply runtime/oti/VMHelpers.hpp:2162]: Note that asserts in header files are problematic since we don't know what module we're in.

Comment (gacholio) [reply runtime/oti/VMHelpers.hpp:2162]: Perhaps these should not be inline functions at all.

**Review by 0xdaryl — APPROVED** [review]

File: runtime/jcl/common/thread.cpp
Comment (gacholio) [reply runtime/jcl/common/thread.cpp:344]: Yes, the use of release without acquire is a timebomb.

File: runtime/vm/BytecodeInterpreter.hpp
Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: `freeContinuation` frees the native memory and NULLs the object field. For concurrent GC, what prevents the invalid use of the native memory pointer (GC thread reads the field as another thread is about to free it)? @LinHu2016

Comment (amicic) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: Concurrent GC will try to atomically win the right to scan by setting a scan bit in continuation state, if (among other conditions) continuation is not finished.
Finished is one of the flags in continuation state and it's set during last unmount before the struct is freed and pointer reset.
On the other side last unmount (setting finished flag in the continuation state) cannot occur while concurrent scan is in progress (in fact even a preceding mount cannot occur, and will wait till scan is complete).

Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: There are no atomics in `freeContinuation` so that code is presumably somewhere in the callers.

Comment (amicic) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: Indeed, freeContinuaiton (or anywhere around unmount path) there are no atomics associated with this synchronization. As mentioned, in the context of carrier thread, it is earlier during mounting that must have occurred before: in synchronizeWithConcurrentGCScan.

In the context of concurrent GC thread, synchronization/atomics occur in tryWinningConcurrentGCScan

However, I'm coming to realization that continuation state should really be a part (a hidden field) of Continuation Java object (btw, I even suggested in past that whole Continuation struct could really be an extension of Java object, mostly for other reasons, reducing stress on native heap etc).
As we have it now, seems like concurrent GC thread doing atomics in tryWinningConcurrentGCScan while racing with last unmount, could operate on the state field that could just been freed (and in very unlikely scenario be reallocated by unrelated party that could initialize that same piece of memory so that it satisfies all the conditions necessary to concurrent scan to win, while it should not).

Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: This really has nothing to do with this PR, I just thought about it when I saw the code. We should probably open a new issue to track the timing hole. I think putting the state in the java object is the only way to address the free timing hole, as you say.

**Comment by gacholio** [comment]
> I also think that acquire/release VMAccess on every iteration we spin is a bit excessive.

This is required when acquiring the inspector for the first time, otherwise we would end up with a situation like:

T1/T2 has VM access
T1 acquires inspector
T2 spins trying to acquire inspector
T1 goes for exclusive
T2 spins forever and we deadlock

**Comment by tajila** [comment]
> This is required when acquiring the inspector for the first time, otherwise we would end up with a situation like:

My point was is it needed everything iteration? Maybe spin a few times then release? My point was really about the cost aqcuire/release each time. I'm stating this assuming that if we don't spin many iterations it may be better to not release at all. I think this is something we should look at and optimize in the future.

**Comment by tajila** [comment]
Yes, I agree. Correctness is the priority now

**Comment by gacholio** [comment]
There's a change or two to be made here, and a rebase/conflict resolve needed before we can proceed.

File: runtime/vm/BytecodeInterpreter.hpp
Comment (LinHu2016) [reply runtime/vm/BytecodeInterpreter.hpp:5283]: I will open new issue about the free timing hole between checking if NULL != continuation and checking continuation->state.

**Comment by gacholio** [comment]
Also, please resolve all conversations that have been addressed so we can know that everyone agrees.

File: runtime/vm/BytecodeInterpreter.hpp
Comment (fengxue-IS) [reply runtime/vm/BytecodeInterpreter.hpp]: freeContinuation change tracked in https://github.com/fengxue-IS/openj9/tree/freeContinuation

File: runtime/oti/VMHelpers.hpp
Comment (fengxue-IS) [reply runtime/oti/VMHelpers.hpp:2264]: enter/exit is currently only used in `javanextvmi.cpp` atm, I will move them into `javanextvmi` as static helper calls

Comment (fengxue-IS) [reply runtime/oti/VMHelpers.hpp:2264]: also per previous investigation, there might be a possibility that -1 would be set when enter is attempted. reason for this is that as soon as a vthread have been yielded (state changed) even before the unmountEnd notification have been called, it may be rescheduled to run on another carrier thread which would try enter critical region. so assert here wouldn't be correct

File: runtime/oti/VMHelpers.hpp
Comment (fengxue-IS) [reply runtime/oti/VMHelpers.hpp:2276]: the assertion is done in the caller, I guess it should be moved inside the call to make it more clear

**Comment by fengxue-IS** [comment]
addressed all review comments, locally compile/tested against JTReg stress test on x86 linux.

File: runtime/oti/VMHelpers.hpp
Comment (tajila) [inline runtime/oti/VMHelpers.hpp:2255]: no real changes in this file

**Review by tajila — APPROVED** [review]

**Comment by fengxue-IS** [comment]
test failure due to assertion added in https://github.com/eclipse-openj9/openj9/pull/16713
```
j9vm.224    *   ** ASSERTION FAILED ** at /home/jenkins/workspace/Build_JDK19_x86-64_linux_Personal/openj9/runtime/vm/ContinuationHelpers.cpp:264: ((VM_VMHelpers::isFinished(continuation->state)))
```
This code is incorrect since when a yielded continuation gets freed by GC due to unreachable, this assertion will fail (though state actually correctly reflected). This assertion made the assumption that unfinished vthread object will be kept alive by the VM's `liveVirtualThreadList` even if it is no longer reachable from user code. FYI @LinHu2016

**Comment by fengxue-IS** [comment]
> jenkins test sanity.functional,extended.functional,sanity.openjdk,sanity.system,extended.system xlinux jdk19

PR testing would need to depends on extension repo changes (something like)

> jenkins test sanity.functional,extended.functional,sanity.openjdk,sanity.system,extended.system xlinux jdk19 depends https://github.com/ibmruntimes/openj9-openjdk-jdk19/pull/83

File: runtime/oti/VMHelpers.hpp
Comment (gacholio) [reply runtime/oti/VMHelpers.hpp:2255]: May as well leave it since the blank line is unnecessary.

**Comment by fengxue-IS** [comment]
pushing fix for assertion, PR test link for reference:
https://openj9-jenkins.osuosl.org/job/Test_openjdk19_j9_sanity.openjdk_x86-64_linux_Personal/9/
https://openj9-jenkins.osuosl.org/job/Test_openjdk19_j9_sanity.functional_x86-64_linux_Personal/88/
https://openj9-jenkins.osuosl.org/job/Test_openjdk19_j9_extended.functional_x86-64_linux_Personal/3/
https://openj9-jenkins.osuosl.org/job/Test_openjdk19_j9_sanity.system_x86-64_linux_Personal/2/
https://openj9-jenkins.osuosl.org/job/Test_openjdk19_j9_extended.system_x86-64_linux_Personal/1/

**Comment by gacholio** [comment]
jenkins test sanity.functional,extended.functional,sanity.openjdk,sanity.system,extended.system xlinux jdk19 depends https://github.com/ibmruntimes/openj9-openjdk-jdk19/pull/83

**Comment by gacholio** [comment]
I won't merge this if the testing passes due to the dependency. Someone who is a committer in both repos will have to do it.

**Comment by tajila** [comment]
Sure Ill do it

**Comment by fengxue-IS** [comment]
build failed due to javadoc compile error, pending fix https://github.com/eclipse-openj9/openj9/pull/17077

**Comment by gacholio** [comment]
jenkins test sanity.functional,extended.functional,sanity.openjdk,sanity.system,extended.system xlinux jdk19 depends https://github.com/ibmruntimes/openj9-openjdk-jdk19/pull/83

**Comment by gacholio** [comment]
Build continues to fail after the doc fix has been merged.

**Comment by gacholio** [comment]
Nevermind, PR had not refreshed.

File: runtime/j9vm/javanextvmi.cpp
Comment (gacholio) [inline runtime/j9vm/javanextvmi.cpp:271]: This assertion is invalid in the GC module. Change `SC` to `MM` (and anywhere  else this might occur in the GC code).

Comment (gacholio) [reply runtime/j9vm/javanextvmi.cpp:271]: Sorry, missed the transition from the GC code to this file. Ignore.

File: runtime/oti/j9nonbuilder.h
Comment (keithc-ca) [inline runtime/oti/j9nonbuilder.h:5840]: I don't see an update to DDR code to deal with the removal of `virtualThreadLinkNextOffset`; as a result, the `!vthreads` command is now broken: #17084 tracks that problem.

**Pull request merged**
